### PR TITLE
Add zstd compression support

### PR DESCRIPTION
**Pull request by martinabeleda**
## Motivation

See #1236

## Solution

Add zstd as a `CompressionEncoding` and as a feature/optional dependency.

New to rust so looking for some feedback on my approach to testing.

### CODE REVIEW

**Review by h33p — COMMENTED** [review]
Not a hyper dev, but approach seems sensible and extensible. I would look into the function visibility comment, though.

File: tonic/src/codec/compression.rs
Comment (mpetri) [inline tonic/src/codec/compression.rs:220]: Wouldn't it be nicer to just mirror the gzip approach and use `zstd::stream::read::{Decoder, Encoder}` instead such that the code can look almost identical:

```
let mut zstd_encoder = Encoder::new(
    &decompressed_buf[0..len],
    // FIXME: support customizing the compression level
    6,
)?;
let mut out_writer = bytes::BufMut::writer(out_buf);
std::io::copy(&mut zstd_encoder, &mut out_writer)?;
```

and decompression:

```
let mut zstd_decoder = Decoder::new(&compressed_buf[0..len])?;
let mut out_writer = bytes::BufMut::writer(out_buf);

std::io::copy(&mut zstd_decoder, &mut out_writer)?;
```

File: tonic/src/codec/compression.rs
Comment (mpetri) [inline tonic/src/codec/compression.rs:42]: shouldn't this  be:

```
    pub(crate) fn into_accept_encoding_header_value(self) -> Option<http::HeaderValue> {
        match (self.is_gzip_enabled(), self.is_zstd_enabled()) {
            (true, false) => Some(http::HeaderValue::from_static("gzip,identity")),
            (false, true) => Some(http::HeaderValue::from_static("zstd,identity")),
            (true, true) => Some(http::HeaderValue::from_static("zstd,gzip,identity")),
            (false, false) => None,
        }
    }
```

**Comment by mpetri** [comment]
@martinabeleda any updates on this CR?

Are you waiting for feedback from the actual maintainers? @LucioFranco would you be willing to accept this change? I think this will lead to substantial performance improvement for especially compressible messages.

**Comment by martinabeleda** [comment]
> @martinabeleda any updates on this CR?
> 
> 
> 
> Are you waiting for feedback from the actual maintainers? @LucioFranco would you be willing to accept this change? I think this will lead to substantial performance improvement for especially compressible messages.
> 
> 
> 
> 

@mpetri I just have a few of your comments to address but other than that I'm waiting for maintainers. I'll have some time to finish the PR later today so can try to reach  @LucioFranco on discord

**Comment by LucioFranco** [comment]
Hi sorry for the delay. I think the comments make sense but I am taking a small break for the next few weeks so the progress has been slow here.

File: tonic/src/codec/compression.rs
Comment (martinabeleda) [reply tonic/src/codec/compression.rs:219]: @mpetri I was able to add your changes for `decompress` in [384e95e](https://github.com/hyperium/tonic/pull/1371/commits/384e95ec35c0cd5d1a1a8e7c8411db9ee29c2973).

For `compress`, that's what I originally tried but `ztd::Encoder` does not implement `std::io::Read`:

```
error[E0277]: the trait bound `zstd::Encoder<'_, &mut [u8]>: std::io::Read` is not satisfied
   --> tonic/src/codec/compression.rs:226:27
    |
226 |             std::io::copy(&mut zstd_encoder, &mut out_writer)?;
    |             ------------- ^^^^^^^^^^^^^^^^^ the trait `std::io::Read` is not implemented for `zstd::Encoder<'_, &mut [u8]>`
    |             |
    |             required by a bound introduced by this call
    |
```

Comment (martinabeleda) [reply tonic/src/codec/compression.rs]: Thanks for the tip! Fixed in [0839324](https://github.com/hyperium/tonic/pull/1371/commits/08393243a2f1e7dc42dd53bf509ee915cb6f43aa)

File: tonic/src/codec/compression.rs
Comment (mpetri) [inline tonic/src/codec/compression.rs:223]: maybe this can be `zstd::DEFAULT_COMPRESSION_LEVEL`?

**Review by mpetri — APPROVED** [review]
Overall this looks good now. One minor comment on the default value of the compressor

**Comment by QuentinPerez** [comment]
Hi @LucioFranco,

I have tested it in production, and it works perfectly! Thank you, @martinabeleda! When do you think we can merge it? I don't want to remain on a fork with that patch forever.

**Comment by LucioFranco** [comment]
@martinabeleda seems like there are some failures around feature flags in CI.

**Comment by QuentinPerez** [comment]
@martinabeleda do you have the time to fix the CI issue ? Otherwise I will re-open the PR with your code and the fixes for the CI.

**Comment by martinabeleda** [comment]
> @martinabeleda do you have the time to fix the CI issue ? Otherwise I will re-open the PR with your code and the fixes for the CI.

@QuentinPerez I haven't had time to work on the fixes so would appreciate if you could help fix CI. Thanks!!

**Comment by tottoto** [comment]
Resolved in #1532. Thank you!

**Pull request closed without merging**
